File: Day3/puzzle-d3-1.pl
use strict;
use warnings;

# Read the file content
sub read_file {
    my $filename = shift;
    open my $fh, '<', $filename or die "Can't open file $filename: $!\n";
    my $content = do { local $/; <$fh> };
    close $fh;
    return $content;
}

# Extract and process valid mul instructions
sub process_instructions {
    my $content = shift;
    my $total = 0;

    # Match valid mul instructions and process them
    while ($content =~ /mul\((\d{1,3}),\s*(\d{1,3})\)/g) {
        my ($num1, $num2) = ($1, $2);
        $total += $num1 * $num2;
    }

    return $total;
}

sub main {
    my $filename = 'input.txt';  
    my $content = read_file($filename);
    my $result = process_instructions($content);
    print "Total sum of valid mul results: $result\n";
}

main();

File: Day3/puzzle-d3-1.py
import re

def process_file(filename):
    try:
        with open(filename, 'r') as file:
            content = file.read()

        # Define a regular expression to match valid mul(X,Y) instructions
        pattern = r"mul\((\d{1,3}),(\d{1,3})\)"

        # Find all matches in the content
        matches = re.findall(pattern, content)

        # Calculate the sum of valid multiplication results
        total_sum = 0
        for match in matches:
            number1, number2 = map(int, match)
            total_sum += number1 * number2

        print(f"The total sum of valid mul instructions is: {total_sum}")

    except FileNotFoundError:
        print(f"Error: The file '{filename}' does not exist.")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    filename = "testinput.txt"
    process_file(filename)